pkg/internal/tests/cross-tests/adapt.go

```diff
@@ -17,13 +17,134 @@
 package crosstests
 
 import (
+	"context"
+	"fmt"
 	"math/big"
 
 	"github.com/hashicorp/terraform-plugin-go/tftypes"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/resource"
 	"github.com/pulumi/pulumi/sdk/v3/go/common/util/contract"
+	"github.com/stretchr/testify/require"
 	"github.com/zclconf/go-cty/cty"
+
+	"github.com/pulumi/pulumi-terraform-bridge/v3/pkg/convert"
+	"github.com/pulumi/pulumi-terraform-bridge/v3/pkg/tfbridge/info"
+	shim "github.com/pulumi/pulumi-terraform-bridge/v3/pkg/tfshim"
+	"github.com/pulumi/pulumi-terraform-bridge/v3/unstable/logging"
 )
 
+// InferPulumiValue is a special marker value that tells cross-tests to infer the Pulumi
+// value from the TF value.
+func InferPulumiValue() resource.PropertyMap {
+	return resource.PropertyMap{
+		inferPulumiValueKey: resource.NewProperty(&inferPulumiValueSecret),
+	}
+}
+
+const inferPulumiValueKey = "inferPulumiValue"
+
+var inferPulumiValueSecret = resource.Secret{}
+
+func shouldInferredPulumiValue(m resource.PropertyMap) bool {
+	v, ok := m[inferPulumiValueKey]
+	return ok && v.IsSecret() && v.SecretValue() == &inferPulumiValueSecret
+}
+
+func inferPulumiValue(t T, schema shim.SchemaMap, infos map[string]*info.Schema, v cty.Value) resource.PropertyMap {
+	if v.IsNull() {
+		return nil
+	}
+	decoder, err := convert.NewObjectDecoder(convert.ObjectSchema{
+		SchemaMap:   schema,
+		SchemaInfos: infos,
+	})
+	require.NoError(t, err)
+
+	ctx := logging.InitLogging(context.Background(), logging.LogOptions{})
+	// There is not yet a way to opt out of marking schema secrets, so the resulting map might have secrets marked.
+	pm, err := convert.DecodePropertyMap(ctx, decoder, ctyToTftypes(v))
```

Comment:
Should we remove this middle step of going to the tftypes?

Reply:
If we have code that does cty.Value to resource.PropertyMap, then yes. I don't want to re-implement this code, since `convert.DecodePropertyMap` is what we use to handle schema information (like renames, maxitemsone, etc.).

Reply:
Yeah, makeTerraformOutput is our conversion function from cty -> propertyMap. Might require some refactoring so feel free to leave as a follow-up